pkg/quarterdeck/auth.go

```diff
@@ -637,6 +639,126 @@ func (s *Server) Refresh(c *gin.Context) {
+	var claims *tokens.Claims
+	if claims, err = middleware.GetClaims(c); err != nil {
+		sentry.Error(c).Err(err).Msg("could not get user claims from authenticated request")
+		c.JSON(http.StatusUnauthorized, api.ErrorResponse(api.ErrInvalidUserClaims))
+		metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "no access token").Inc()
+		return
+	}
+
+	// If the orgID in the claims is the same as the requested orgID return an error
+	// (the user must use the refresh endpoint to get claims for the same orgID)
+	if orgID := claims.ParseOrgID(); orgID.Compare(in.OrgID) == 0 {
+		sentry.Warn(c).Msg("user attempting to switch to the same organization")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("cannot switch into the organization you are currently logged into"))
+		metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "switch to same org").Inc()
+		return
+	}
+
+	// Verify that the user belongs to the requested organization.
+	var orgUser *models.OrganizationUser
+	if orgUser, err = models.GetOrgUser(c.Request.Context(), claims.Subject, in.OrgID); err != nil {
+		if errors.Is(err, db.ErrNotFound) {
+			sentry.Warn(c).Str("userID", claims.Subject).Str("orgID", in.OrgID.String()).Msg("user attempt to switch into organization they do not belong to")
+			c.JSON(http.StatusForbidden, api.ErrorResponse("invalid credentials"))
+			metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "invalid credentials").Inc()
+			return
+		}
+
+		sentry.Error(c).Err(err).Msg("could not retrieve organization user from database")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not process switch request"))
+		metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "unhandled error").Inc()
+		return
+	}
+
+	// Fetch the user information and create claims now that they are successfully authenticated
+	var user *models.User
+	if user, err = orgUser.User(c.Request.Context(), false); err != nil {
+		sentry.Error(c).Err(err).Msg("could not retrieve user from database")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not process switch request"))
+		metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "unhandled error").Inc()
+		return
+	}
+
+	// Create access and refresh tokens for new organization
+	newClaims := &tokens.Claims{
+		RegisteredClaims: jwt.RegisteredClaims{
+			Subject: user.ID.String(),
+		},
+		Name:    user.Name,
+		Email:   user.Email,
+		Picture: gravatar.New(user.Email, nil),
+		OrgID:   orgUser.OrgID.String(),
+	}
+
+	// Add the user permissions to the claims
+	if claims.Permissions, err = user.Permissions(c.Request.Context(), false); err != nil {
+		sentry.Error(c).Err(err).Msg("could not get user permissions from model")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not create credentials"))
+
+		// increment failure count
+		metrics.FailedLogins.WithLabelValues(ServiceName, UserHuman, "permissions not found").Inc()
+		return
+	}
```

Comment:
It's _really_ important that new claims are created and that the old claims are NOT reused; should I stress that more in comments here; should I make it more secure so that we don't accidentally put in a bug with the old claims?

Reply:
Am I missing something or is there already a bug? Don't the permissions need to be copied into `newClaims`? Maybe one way to make this more foolproof is to create a method on the `User` which constructs the new claims.

Reply:
The bug would be if anything was copied from the old claims to the new claims. The permissions must come from the new organization role; e.g. if the user is an Owner in Organization A but a member in Organization B; the new claims must reflect the role in B if they're switching from A to B; replacing the claims from A.